spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/DataFileServices.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.iceberg.spark.data.SparkParquetReaders;
+import org.apache.iceberg.spark.data.SparkParquetWriters;
+import org.apache.iceberg.spark.data.SparkPlannedAvroReader;
+import org.apache.iceberg.spark.data.vectorized.VectorizedSparkOrcReaders;
+import org.apache.iceberg.spark.data.vectorized.VectorizedSparkParquetReaders;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.vectorized.ColumnarBatch;
+import org.apache.spark.unsafe.types.UTF8String;
+
+public class DataFileServices {
+  public static void register() {
+    // Base readers
+    DataFileServiceRegistry.registerReader(
+        FileFormat.PARQUET,
+        InternalRow.class.getName(),
+        inputFile -> Parquet.read(inputFile).readerFunction(SparkParquetReaders::buildReader));
+
+    DataFileServiceRegistry.registerReader(
+        FileFormat.AVRO,
+        InternalRow.class.getName(),
+        inputFile -> Avro.read(inputFile).readerFunction(SparkPlannedAvroReader::create));
+
+    DataFileServiceRegistry.registerReader(
+        FileFormat.ORC,
+        InternalRow.class.getName(),
+        inputFile -> ORC.read(inputFile).readerFunction(SparkOrcReader::new));
+
+    // Vectorized readers
+    DataFileServiceRegistry.registerReader(
+        FileFormat.PARQUET,
+        ColumnarBatch.class.getName(),
+        ParquetReaderType.ICEBERG.name(),
+        inputFile ->
+            Parquet.read(inputFile)
+                .batchReaderFunction(
+                    (schema, messageType, idToConstant, deleteFilter) ->
+                        VectorizedSparkParquetReaders.buildReader(
+                            schema,
+                            messageType,
+                            idToConstant,
+                            (DeleteFilter<InternalRow>) deleteFilter)));
+
+    DataFileServiceRegistry.registerReader(
+        FileFormat.PARQUET,
+        ColumnarBatch.class.getName(),
+        ParquetReaderType.COMET.name(),
+        inputFile ->
+            Parquet.read(inputFile)
+                .batchReaderFunction(
+                    (schema, messageType, idToConstant, deleteFilter) ->
+                        VectorizedSparkParquetReaders.buildCometReader(
+                            schema,
+                            messageType,
+                            idToConstant,
+                            (DeleteFilter<InternalRow>) deleteFilter)));
+
+    DataFileServiceRegistry.registerReader(
+        FileFormat.ORC,
+        ColumnarBatch.class.getName(),
+        inputFile ->
+            ORC.read(inputFile).batchReaderFunction(VectorizedSparkOrcReaders::buildReader));
+
+    DataFileServiceRegistry.registerAppender(
+        FileFormat.AVRO,
+        InternalRow.class.getName(),
+        outputFile ->
+            Avro.write(outputFile)
+                .writerFunction(
+                    (schema, engineSchema) -> new SparkAvroWriter((StructType) engineSchema))
```

Comment:
Why is there a cast here if the builder is parameterized by this type?

Reply:
Had to make Avro.WriteBuilder parametrized. Created a deprecation and new classes to make this work. Added serious size to the PR.
Please check, and if you think it is worth it, then we can keep